signature/src/balrog.rs

```diff
@@ -0,0 +1,382 @@
+            let subject = &self.chain[index];
+            let issuer = &self.chain[index+1];
+            let _ = match Self::verify_cert_chain_pair(subject, issuer, current_time) {
+                Err(e) => return Err(e),
+                Ok(x) => x
+            };
+            index = index+1;
+        }
+
+        /* Verify the root certificate. */
+        let root_cert = self.chain.last().unwrap();
+        let _ = match Self::verify_cert_chain_pair(root_cert, root_cert, current_time) {
+            Err(e) => return Err(e),
+            Ok(x) => x
+        };
+
+        let hash_stripped = root_hash.replace(":", "");
+        let hash_decoded = match hex::decode(hash_stripped) {
+            Err(_e) => return Err(BalrogError::RootHashParseFailed),
+            Ok(x) => x
+        };
+        if self.root_hash != hash_decoded {
+            return Err(BalrogError::RootHashMismatch);
+        }
+        
+        /* Success! */
+        Ok(())
+    }
+
+    /* Internal helper function to check that a certificate was signed by its issuer. */
+    fn verify_cert_chain_pair(
+        subject: &X509Certificate,
+        issuer: &X509Certificate,
+        current_time: i64,
+    ) -> Result<(), BalrogError> {
+        /* Check that the subject and issuer names match. */
+        if subject.issuer() != issuer.subject() {
+            return Err(BalrogError::ChainSubjectMismatch);
+        }
+
+        /* Check the certificate validity period. */
+        let asn_timestamp = ASN1Time::from_timestamp(current_time).unwrap();
+        if !subject.validity().is_valid_at(asn_timestamp) {
+            return Err(BalrogError::CertificateExpired);
+        }
+        if !issuer.validity().is_valid_at(asn_timestamp) {
+            return Err(BalrogError::CertificateExpired);
+        }
+
+        /* Check that the issuer is permitted to sign certificates. */
+        let ca_key_usage = match issuer.key_usage() {
+            Err(e) => return Err(BalrogError::from(e)),
+            Ok(x) => x
+        };
+        let ca_key_usage = ca_key_usage.unwrap().value;
+        if !ca_key_usage.key_cert_sign() {
+            return Err(BalrogError::IssuerUnauthorized);
+        }
+        if !issuer.is_ca() {
+            return Err(BalrogError::IssuerUnauthorized);
+        }
+
+        /* Check X509 certificate validation. */
+        let mut logger = VecLogger::default();
+        // structure validation status
+        let ok = X509StructureValidator
+            .chain(X509CertificateValidator)
+            .validate(subject, &mut logger);
+        if !ok {
+            let reason = match logger.errors().first() {
+                None => "Unknown error",
+                Some(s) => s
+            };
+            return Err(BalrogError::X509ValidationError(reason.to_string()));
+        }
+        
+        /* Verify the cryptographic signature. */
+        subject.verify_signature(Some(issuer.public_key()))
+            .map_err(|e| { BalrogError::from(e) })
```

Comment:
This ones can also be replaced by a `?`.

Reply:
Will do!